fdbcli/DebugCommands.actor.cpp

```diff
@@ -368,19 +321,68 @@ ACTOR Future<bool> checkallCommandActor(Database cx, std::vector<StringRef> toke
 						replies.push_back(s.getKeyValues.getReplyUnlessFailedFor(req, 2, 0));
 					}
 					wait(waitForAll(replies));
-					auto res = checkResults(version, keyServers[i].second, replies);
-					bool allSame = std::get<0>(res);
-					hasMore = std::get<1>(res);
-					Key nextBeginKey = std::get<2>(res);
-					// Using claimEndKey of the current round as the nextBeginKey for the next round
-					// Note that claimEndKey is not compared in the current round
-					// This key will be compared in the next round
-					printf("Result: compared %s - %s\n",
-					       printable(beginKeyToCheck.toString()).c_str(),
-					       printable(nextBeginKey.toString()).c_str());
-					beginKeyToCheck = nextBeginKey;
-					printf("allSame %d, hasMore %d, checkAll %d\n", allSame, hasMore, checkAll);
-					if (!allSame && !checkAll) {
+
+					// Decide comparison scope
+					Key claimEndKey; // used for the next round if hasMore == true
+					Key maxEndKey;
+					hasMore = false; // re-calculate hasMore according to replies
+					for (int j = 0; j < replies.size(); j++) {
+						auto reply = replies[j].get();
+						if (reply.isError()) {
+							printf("checkResults error: %s\n", reply.getError().what());
+							throw reply.getError();
+						} else if (reply.get().error.present()) {
+							printf("checkResults error: %s\n", reply.get().error.get().what());
+							throw reply.get().error.get();
+						}
+						GetKeyValuesReply current = reply.get();
+						if (current.data.size() == 0) {
+							continue; // Ignore if no data has replied
+						}
+						if (claimEndKey.empty() || current.data[current.data.size() - 1].key < claimEndKey) {
+							claimEndKey = current.data[current.data.size() - 1].key;
+						}
+						if (maxEndKey.empty() || current.data[current.data.size() - 1].key > maxEndKey) {
+							maxEndKey = current.data[current.data.size() - 1].key;
+						}
+						hasMore = hasMore || current.more;
+					}
+					printf("Compare scope has been decided\n\tBeginKey: %s\n\tEndKey: %s\n\tHasMore: %d\n",
+					       printable(beginKeyToCheck).c_str(),
+					       printable(claimEndKey).c_str(),
+					       hasMore);
+					if (claimEndKey.empty()) {
+						printf("checkResults error: claimEndKey is empty\nPlease re-run the checkall.\n");
+						throw operation_failed();
+					} else if ((beginKeyToCheck == claimEndKey) && hasMore) {
```

Comment:
IIUC, its possible that the range is [1, 10], while SS1 has [2,3,4], and SS2 has [2], and this would not work. So we need to compare current.data[0] and current.data.[size() - 1], rather than the range begin.

Reply:
If SS1 has [2,3,4], and SS2 has [2], then the process will enter the third case, and completed by the third case.